### ISSUE TITLE
DSFT 01 이재우 데이터 분석 무작정 따라하기 #1

### ISSUE THREAD

**Issue by jingwoo4710** [codestates/ds-blog#59]
# 1. 데이터 분석 무작정 따라하기
학생이라면 아니 아마 대부분의 사람들은 공부를 어떻게 해야하나? 라는 질문을 안고 살고 있다. 내가 고등학교 3학년일 때, 지옥같던 하루 들을 참아내면서 공부하며 꿈을 꾸던 공부 보단 자유를 가진 대학생, 성인이 아닌 여전히 공부하고 있는 지금의 내 모습만 봐도 알 수 있다. 그렇기 때문에, 인생은 공부의 연속이라는 말에 크게 공감할 수 있다. 공부를 어떻게 하면 잘 할 수 있을까? 지금까지 공부를 했지만, 잘하는 방법은 너무나 다양하고 사람마다 다르기 때문에 정답은 없다. 하지만, 나는 데이터 사이언스를 공부하면서 내가 생각들을 이해한 부분들을 나만의 방식으로 이해하고 기록하고 정리해서 내 입맛에 맞는 나의 말로 이해한 책을 만들것이다.

##   &nbsp; &nbsp;&nbsp; &nbsp;   1.1 대상
&nbsp; &nbsp;  &nbsp; &nbsp; &nbsp; [Recommender System with Python](https://lsjsj92.tistory.com/597)

# 2. 방법
- 코드 이해하고, 나의 말로 바꿔보기
-  [Youtube](https://www.youtube.com/channel/UCkmF8xvMW1K8_Xbvc1JcrbQ) 에 매일 조금이라도 공부하는 영상 올리기
- [Blog](https://github.com/codestates/ds-blog/issues) 에 기록하기

# 3. 데이터 전처리 코드 분석
```py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from ast import literal_eval
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.metrics.pairwise import cosine_similarity

```
> 항상 이 과정을 왜 제일 위에 쓸까 생각했는데, 이건 이 코드를 따라하고 읽는 사람을 위한 배려라고 느꼈다. 막상 따라할려고 코드를 Run 했는데 이것들이 없어 생긱는 에러를 만날 생각하니 까마득하다. 코드는 읽는 사람을 배려해줘야 하는것 같다. 나만 잘해서는 그 누구도 알아줄 수 없다.

💡 import 는 전체를 가져오는 것, from X import Y 는 X 에서 필요한 Y 만 가져오는 것.
```py
data = pd.read_csv('./movie_data/tmdb_5000_movies.csv')
```
> **pd.read_csv()** : 데이터 파일은 csv 파일로 많이 저장되고, csv 파일을 읽는 메서드 
👨  tmdb_5000_movies.csv 라는 파일을 data로 저장할거야.

```py
data.head(2)
```
> **Dataframe.head(2)** : 데이터의 가장 위 2개를 보여줌, 데이터가 어떻게 생겼는지 간략히 확인
 👨 상위 2개의 데이터를 보고싶어.

```py
data.shape 
```
> **Dataframe.shape** : 매트릭스 (row, column) 을 보여줌, 데이터의 큰 포맷을 확인
 👨 포맷이 어떻게 되는지 보고싶어, 데이터의 크기를 대략적으로 보고싶어.

```py
data = data[['id','genres', 'vote_average', 'vote_count','popularity','title',  'keywords', 'overview']]
```
>  **Dataframe[["col_name1" , "col_name2"]]** : 불 필요한 열을 제거하는 과정!
👨 data에서 'id', 'genres', 'vote_average', 'vote_count', 'popularity','title',  'keywords', 'overview' 만 볼거야.

```py
m = data['vote_count'].quantile(0.9)
```
> **Series.quantile()** : 상위 몇퍼센트의 값을 알고 싶을 때
👨 data에서 'vote_count'에서 상위 10%의 값을 알고 싶고 그 값을 m 에 저장할거야.
```py
data = data.loc[data['vote_count'] >= m]
```
> **Dataframe.copy()** : Data의 원본을 해치지 않기 위해서 복사하기
**Dataframe.copy()**.loc[data['col_name'] >= value] : 특정 열에서 특정 값이 이상인 데이터만 추출하기
👨 data를 수정할건데, 이 데이터는 data에서 'vote_count' 열의 값이 m 보다 같거나 큰 데이터만 모아서 만들거야.
```py
 C = data['vote_average'].mean()
```
> **Series.mean()** : 데이터의 한 열의 평균값
👨 c 라는 변수에, data의 'vote_average' 라는 열의 평균값을 저장할거야.
```py
def weighted_rating(x, m=m, C=C):
    v = x['vote_count']
    R = x['vote_average']
    
    return ( v / (v+m) * R ) + (m / (m + v) * C)
```
>👨  weighted_rating 이라는 함수를 정의 할건데, x는 우리가 넘겨줘야 하는 parameter 이고, m과 C 미리 선언해서 값을 미리 정해놨어. x는 Dataframe일 것이고, v는 그 데이터에서 'vote_count'라는 값을 저장하고 R은 'vote_average' 값을 저장할거야. 그리고 함수를 호출 했을때 결과값은 ( v / (v+m) * R ) + (m / (m + v) * C) 이야. 
❓ v와 R은 Series 타입일텐데, int인 m과 C를 함께 수식을 짤 수 있구나. Element-wise 계산인가?

```py
data['score'] = data.apply(weighted_rating, axis = 1)
```
>**Dataframe.apply()** : 특정 함수를 data에 적용하는 메서드, axis=1은 each row에 적용.
👨 data에 'score'라는 열을 추가 할건데, each row에 weighted_rating이라는 함수를 통해서 계산된 값을 저장해서 추가 할거야.

💡 데이터에서 한 row 에서 다른 col의 값을 이용하여 계산을 하려면 값을 계산하는 함수를 저장하고, apply(함수, axis=1)를 통해서 가능하구나.
```py
data['genres'] = data['genres'].apply(literal_eval)
data['keywords'] = data['keywords'].apply(literal_eval)
```
> ast.literal_eval : 문자열에 있는 문구를 문자열 때고 evaluate 해주는 메서드 ex) '{'id':28}' (string)  👉  {'id':28} (dictionary)
data의 'genres'와 'keywords' 열이 string으로 되어있는데, string을 떼고 그 값을 읽어서 저장하고 싶어.

```py
data['genres'] = data['genres'].apply(lambda x : [d['name'] for d in x]).apply(lambda x : " ".join(x))
data['keywords'] = data['keywords'].apply(lambda x : [d['name'] for d in x]).apply(lambda x : " ".join(x))
```
> **lambda x : return**  👉  정의 하지 않은 함수를 사용하고 싶을 때 
**list." ".join()** : 특정 list의 값이 string으로 되어있는데, 그 string을 space 로 구분해서 하나의 string으로 저장
👨 data의 'genres'와 'keywords'의 값은 dictionary 타입이다. 그 dictionary 값 중에서 특정 'key'에 해당하는 값만 사용하고 싶다. 그래서
data의 'genres'와 'keywords'의 값을 변형하고 싶은데, data의 'genres'와 'keywords' 열의 값에서 'name'이라는 key의 값을 list로 저장하고 저장 된 list에 element 인 string들을 space로 구분하여 하나의 string으로 data 'genres' 와 'keywords' 열에 저장!

💡 apply와 lambda를 통해서 다양한 방식으로 데이터 값 변경이 가능하구나!
💡 데이터 값이 dictionary 양식일 때, 특정 key의 값만 필터링 하는 방법
```py
data.to_csv('./movie_data/pre_tmdb_5000_movies.csv', index = False)
```
> **Dataframe.to_csv()** : Dataframe을 csv 파일로 저장 
👨 data라는 데이터를 ./movie_data/pre_tmdb_5000_movies.csv 으로 인덱스 없이 저장

# 4. Titanic Data에 적용해보기
[Titanic_1.ipynb](https://colab.research.google.com/drive/1F50nxrBH1drOVpQ15K6b5zciPb2TPsSk)

**Comment by seungtaemoon** [comment]
잘 읽었습니다.

재우님은 전공자이신가요? 전의 advanced challenge 하신 걸 읽어보니 다른 분들보다 조금 더 이 분야에 대해 이해도가 높은 분 처럼 느껴집니다.

좋은 설명감사합니다! 앞으로도 쓰시는 글 기대하겠습니다~!

**Comment by jingwoo4710** [comment]
답변 감사합니다.
어느 과를 전공해야 전공자라고 말할 수 있나 싶은데, 제 전공은 Econometric 이며 어느 정도 Data Scienc 와 닮아 있다고 생각합니다. 
항상 댓글 달아주셔서 감사합니다. 글 쓰는것에 대해서 저도 승태님께 배울점을 많이 보여저서요, 저도 열심히 배우겠습니다.
감사합니다.

**Comment by jhk0530** [comment]
재우님, 저는 일단 올리신 글이 오늘의 *과제*로 제출 하신 것 같지는 않다고 생각이 되지만, 
만약 이 글이 과제로 제출하신 내용이라면 이 과제에 대해서 점수를 0점을 드릴겁니다. 

이유는 다음과 같습니다. 
아시겠지만, 이번 과제는 5개의 주제
- git, github
- python environment
- data visualization
- titanic data analysis
- python class oop

중 자신이 원하는 한가지 주제를 선택하여, 그 주제를 바탕으로 하여, 
**독자를 고려한 블로그 글**을 작성하는 것이었습니다.

그러나 제가 **올려주신 과제** 및 포함되어 있는 **노트북**과 **유튜브**를 확인했을 때, 재우님께서 올리신 내용은 

위 5가지 주제 중 어느 것에도 해당 하지 않는다고 판단했습니다.

독자가 이 글을 읽었을때 타이타닉 데이터에 대한 인사이트를 얻을 수 있었나? 라는 질문에 답을 하기도 좀 어려울 것 같구요.

시각화나 python 혹은 git 에 대한 내용은 포함 되어 있지 않다고 봐도 무방 할 것 같아요.

물론 코드마다 디테일 하게 어떠한 기능을 하는지 설명해 주신 노력은 대단하다고 생각합니다. 저는 잘 하지 못하는 부분이라서 부럽기도 합니다.

하지만, Data Analysis 과정에서는 해당 코드의 작동 방식을 설명하는 것도 물론 중요합니다. 중요한데
왜 이러한 operation을 했는지 를 독자에게 설명하는 것이 더욱 중요하다고 생각해요.

가령 

`data = data[['id','genres', 'vote_average', 'vote_count','popularity','title',  'keywords', 'overview']]`

부분에서 불필요한 column을 제거하는 역할을 한다고 설명을 하신건 아주 좋습니다. 

10점 만점에 추가로 1점 가산점을 받을 수 있는 행동이라고 생각해요. 

그러나 **해당 column들이 왜 불필요한지, 왜 제거해야 하는지** 가 설명되지 않으면 이 추가 점수는 의미가 없어집니다.

---

저는 재우님의 속마음이나 의도를 알 수 없어서 나름의 생각을 해본 결과. 

타이타닉 데이터 분석을 하고 싶었는데, 분석에 대한 인사이트를 처음 `대상`부분에 올려주신 

블로그의 내용에서 제시하는 단계들을 바탕으로 따라서 분석을 했다. 

라는 의도를 갖지 않으셨을까 조심스럽게 생각해봅니다. 

만약에 제 생각과 비슷한 의도를 가지고 작업을 하셨다면, 해당 내용을 언급해주셔야. 

저 혹은 다른 수강생들을 비롯한 *독자*가 재우님께서 시간을 들여서 쓰신 글을 더 쉽게 이해 할 수 있다고 생각해요.

---

재우님이 저희 코스를 수료하고, 취업하게 될 Data Scientist / Analyst 는 다양한 배경을 가진 사람들과 다양한 내용의 업무를 접하게 될 겁니다. 

테크니컬한 요소도 중요하지만 커뮤니케이션 또한 정말 필요한 요소라고 생각해요.

그리고 아마 글로써 표현하는게 익숙하지 않으셔서 제가 재우님의 의도를 오해했다고 저는 생각하고 있습니다.

채점은 오늘 밤까지 해야 하지만, 재우님께서 (가능하면 내일 오전 이전에) 제가 궁금해 했던 이슈들에 대해서 답변을 해주실때까지 잠시 보류하도록 
하겠습니다.

이번 첫 주 저희와 같이 노력해주셔서 감사합니다. 평안한 밤 되세요.

**Comment by jingwoo4710** [comment]
피드백 감사합니다.
말씀하신 대로 주어진 주제에 맞지 않는 과제로 0점을 받았어야 합니다. 기회를 주셔서 감사합니다.
너무 작은것들에 신경을 쓰다보니 큰 그림이 틀린것 같습니다. 제가 글을 쓰고자 하는 의도와 그걸 읽는 사람이 좀 더 잘 이해할 수 있도록 일단 바꾸긴 했습니다.
내일 오전중으로 라고 말씀하셨는데, 정확한 시간을 알려주시면 최종점검 하고 다시 올려볼 수 있도록 하겠습니다.
감사합니다.
